File: app/src/main/java/com/ed/redditapp/ui/search/SearchDaggerModule.java
package com.ed.redditapp.ui.search;

import com.ed.redditapp.FragmentScope;
import com.ed.redditapp.lib.api.RedditApi;

import dagger.Module;
import dagger.Provides;

@Module
public class SearchDaggerModule {
    private final SearchFragment fragment;

    public SearchDaggerModule(SearchFragment fragment) {
        this.fragment = fragment;
    }

    @Provides
    @FragmentScope
    public SearchModel provideModel(RedditApi redditApi) {
        return new SearchModel(redditApi);
    }

    @Provides
    @FragmentScope
    public SearchPresenter providePresenter(SearchModel model) {
        return new SearchPresenter(model, fragment);
    }

    @Provides
    @FragmentScope
    SearchAdapter provideAdapter(SearchPresenter presenter) {
        return new SearchAdapter(fragment.getActivity(), presenter);
    }
}

File: app/src/main/java/com/ed/redditapp/ui/search/SearchDaggerModule.kt
package com.ed.redditapp.ui.search

import com.ed.redditapp.FragmentScope
import com.ed.redditapp.lib.api.RedditApi
import dagger.Module
import dagger.Provides

@Module
class SearchDaggerModule(private val fragment: SearchFragment) {
    @Provides @FragmentScope
    fun provideModel(redditApi: RedditApi) = SearchModel(redditApi)

    @Provides @FragmentScope
    fun providePresenter(model: SearchModel) =
        SearchPresenter(model, fragment)

    @Provides @FragmentScope
    fun provideAdapter(presenter: SearchPresenter) =
        SearchAdapter(fragment.requireContext(), presenter)
}